src/test/firestore/indexes.spec.ts

```diff
@@ -0,0 +1,221 @@
+import { expect } from "chai";
+import { FirestoreIndexes } from "../../firestore/indexes";
+import * as API from "../../firestore/indexes-api";
+import * as Spec from "../../firestore/indexes-spec";
+
+const idx = new FirestoreIndexes();
+
+const VALID_SPEC = {
+  indexes: [
+    {
+      collectionGroup: "collection",
+      queryScope: "COLLECTION",
+      fields: [
+        { fieldPath: "foo", order: "ASCENDING" },
+        { fieldPath: "bar", order: "DESCENDING" },
+        { fieldPath: "baz", arrayConfig: "CONTAINS" },
+      ],
+    },
+  ],
+  fieldOverrides: [
+    {
+      collectionGroup: "collection",
+      fieldPath: "foo",
+      indexes: [
+        { order: "ASCENDING", scope: "COLLECTION" },
+        { arrayConfig: "CONTAINS", scope: "COLLECTION" },
+      ],
+    },
+  ],
+};
+
+describe("IndexValidation", () => {
+  it("should accept a valid v1beta2 index spec", () => {
+    idx.validateSpec(VALID_SPEC);
+  });
+
+  it("should not change a valid v1beta2 index spec after upgrade", () => {
+    const upgraded = idx.upgradeOldSpec(VALID_SPEC);
+    expect(upgraded).to.eql(VALID_SPEC);
+  });
+
+  it("should accept a valid v1beta1 index spec after upgrade", () => {
+    idx.validateSpec(
+      idx.upgradeOldSpec({
+        indexes: [
+          {
+            collectionId: "collection",
+            fields: [
+              { fieldPath: "foo", mode: "ASCENDING" },
+              { fieldPath: "bar", mode: "DESCENDING" },
+              { fieldPath: "baz", mode: "ARRAY_CONTAINS" },
+            ],
+          },
+        ],
+      })
+    );
+  });
+
+  it("should reject an incomplete index spec", () => {
+    expect(() => {
+      idx.validateSpec({
+        indexes: [
+          {
+            collectionGroup: "collection",
+            fields: [
+              { fieldPath: "foo", order: "ASCENDING" },
+              { fieldPath: "bar", order: "DESCENDING" },
+            ],
+          },
+        ],
+      });
+    }).to.throw();
```

Comment:
`to.throw` can take a error type (e.g. `FirebaseError`) and a regex to match against the message which is very useful to make sure you're getting the _correct_ error.